scripts/sanitycheck

```diff
@@ -321,58 +327,72 @@ class NativeHandler(Handler):
+            verbose("NATIVE: {0}".format(line.decode('utf-8').rstrip()))
+            log_out_fp.write(line.decode('utf-8'))
+            log_out_fp.flush()
+            harness.handle(line.decode('utf-8').rstrip())
+            if harness.state:
+                proc.terminate()
+                break
+
+        log_out_fp.close()
+
     def handle(self):
         out_state = "failed"
 
-        with open(self.run_log, "wt") as rl, open(self.valgrind_log, "wt") as vl:
-            try:
-                binary = os.path.join(self.outdir, "zephyr", "zephyr.exe")
-                command = [binary]
-                if shutil.which("valgrind") and self.valgrind:
-                    command = ["valgrind", "--error-exitcode=2",
-                               "--leak-check=full"] + command
-                returncode = subprocess.call(command, timeout=self.timeout,
-                                             stdout=rl, stderr=vl)
-                self.returncode = returncode
-                if returncode != 0:
-                    if self.returncode == 1:
-                        out_state = "failed"
-                    else:
-                        out_state = "failed valgrind"
-
-            except subprocess.TimeoutExpired:
+        harness_name = self.instance.test.harness.capitalize()
+        harness_import = HarnessImporter(harness_name)
+        harness = harness_import.instance
+        harness.configure(self.instance)
+
+        binary = os.path.join(self.outdir, "zephyr", "zephyr.exe")
+        command = [binary]
+        if shutil.which("valgrind") and self.valgrind:
+            command = ["valgrind", "--error-exitcode=2",
+                       "--leak-check=full"] + command
+
+        with subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE) as proc:
+            t = threading.Thread(target=self._output_reader, args=(proc, harness, ))
+            t.start()
+            t.join(self.timeout)
+            if t.is_alive():
+                proc.terminate()
                 out_state = "timeout"
-                self.returncode = 1
+                t.join()
 
-        with open(self.run_log, "r") as rl:
-            for line in rl.readlines():
-                line = line.strip()
-                if self.RUN_PASSED in line:
-                    out_state = "passed"
-                    break
-
-                if self.RUN_FAILED in line:
+            proc.wait()
+            self.returncode = proc.returncode
+            if proc.returncode != 0:
+                if self.returncode == 1:
                     out_state = "failed"
-                    break
+                else:
+                    out_state = "failed valgrind"
 
-        returncode = subprocess.call(["GCOV_PREFIX=" + self.outdir, "gcov", self.sourcedir, "-s", self.outdir], shell=True)
+        returncode = subprocess.call(["GCOV_PREFIX=" + self.outdir, "gcov", self.sourcedir, "-b", "-s", self.outdir], shell=True)
```

Comment:
Just a note: Most sample apps, are configured without COVERAGE and therefore will not account for this yet.
I guess when/if https://github.com/zephyrproject-rtos/zephyr/issues/5723 is implemented, we could use it to override their configuration. So that when run in sanitycheck, all tests and samples run in the POSIX arch would be compiled with coverage, instead of depending on the project configuration.